### PR TITLE
Implement a driver for the N4AIA04.

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

**Comment by floitsch** [comment]
Currently the PR depends on unpublished packages.
Once they are published, and before committing, I need to change them to the public versions.

File: src/n4aia04.toit [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+  static BAUD_RATE_ ::= 0x0F
+  static BAUD_RATE_1200_ ::= 0
+  static BAUD_RATE_2400_ ::= 1
+  static BAUD_RATE_4800_ ::= 2
+  static BAUD_RATE_9600_ ::= 3
+  static BAUD_RATE_19200_ ::= 4
+
+  registers_/modbus.HoldingRegisters
+
+  constructor station/modbus.Station:
```
Comment (floitsch) [inline src/n4aia04.toit:43]: Document the constructors.

**Comment by gregger2k** [comment]
in test_esp32.toit.... your current calculations are in error.
The 4-20mA signaling spec expects a current source, not a voltage source. The loop will put out whatever voltage is required to output the required current.
You have to add the 200 ohm internal loop resistance to your 330 ohm load to get the total resistance to get the loop current.. so 5v / (200 + 330) ohms = 9.4mA not 15ma. And the 3.3v / 530 ohms will give 6mA ... no calibration adjustment needed. If you multiply these number by your 1.64 magics number the results returned are close to actual.

If you reduce your 330 ohm resistor by the internal loop resistance your returned values will be closer to the 10ma and 15ma you are looking for.. You can check it with a meter or a 4-20 mA loop tester.

**Comment by floitsch** [comment]
Thanks.
I was expecting something like that. The test file even talks about the 200 Ohm register (line 26). I was just assuming that the sensor shouldn't do that. (Was hoping it would behave like my multimeter which basically adds no resistance).

Is the 200 Ohm loop resistance mentioned somewhere in the spec?
From your feedback it seems like this is a common thing, so I will rewrite the comments in the test (and update the values).

**Comment by gregger2k** [comment]
The loop resistance can vary  with the sensor. The sensor usually specifies  maximum loop resistance. You can have multiple current sensing devices in line as long as you do not exceed the maximum loop resistance. The measuring device is just measuring the voltage across the sensing resistor.

Here is one type of signal generator for simulating a 4-20 sensor.
https://www.amazon.com/dp/B0991ZSBHJ?psc=1&ref=ppx_yo2ov_dt_b_product_details

File: src/n4aia04.toit [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+    registers_ = station.holding_registers
+
```
Comment (kasperl) [inline src/n4aia04.toit:45]: Drop one newline.

File: src/n4aia04.toit [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+    register_value /int := ?
+    if baud_rate == 1200: register_value = BAUD_RATE_1200_
+    else if baud_rate == 2400: register_value = BAUD_RATE_2400_
+    else if baud_rate == 4800: register_value = BAUD_RATE_4800_
+    else if baud_rate == 9600: register_value = BAUD_RATE_9600_
+    else if baud_rate == 19200: register_value = BAUD_RATE_19200_
+    else: throw "INVALID_ARGUMENT"
+    write_ --address=BAUD_RATE_ register_value
+
+
```
Comment (kasperl) [inline src/n4aia04.toit]: Drop a couple of newlines.

**Review by kasperl — APPROVED** [review]
Looks good to me.

File: src/n4aia04.toit (hunk 2 above)
Comment (floitsch) [reply src/n4aia04.toit:45]: done.

File: src/n4aia04.toit (hunk 3 above)
Comment (floitsch) [reply src/n4aia04.toit]: done.

File: src/n4aia04.toit (hunk 1 above)
Comment (floitsch) [reply src/n4aia04.toit:43]: done.

**Pull request merged**